### PR TITLE
only save case attachment if content differs

### PR DESCRIPTION
**Pull request by dannyroberts**
Open for tests. I think I want to write another test specifically for this case though.

### CODE REVIEW

**Comment by czue** [comment]
:+1: when test is added. does this also mean that we pull all case attachments from the DB on form processing even if they haven't been touched?

**Comment by dannyroberts** [comment]
Yeah, that's a good point. I took a quick look, and that looks little trickier just given the way it's implemented. Just a recap of my current understanding of the problem:

1. When you submit a case, if any actions are deemed out of order, it rebuilds them all by reapplying all actions on top of the current state
2. When applying an attachment action, fetches the attachment from the form and adds it to the case. <-- this is the part I optimized; it now only adds it if there'd be a change, otherwise it leaves what's already there
3. If there are a few of these as a result of the same form submission, the total bulk save request size can easily exceed 64MB

I'm not 100% about the mechanism by which these particular case actions are getting rebuilt, so (1) here is just a guess.

There is also a potential weakness in this optimization that I see, which is that while it is definitely a strict improvement, it can still miss the case where you attach image A, and then image B. Since at the beginning it starts out as image B, both of those actions would detect a change, and thus we'd end up re-saving image B. The only reasonable solution to this that I see is to all of the planning for applying an action first, and then at the end, decide whether we need to pull/save any attachments. This is a bigger refactor that has potential to effect more things (like when/where errors get raised and handled in the case of bad refs, among other things).

**Comment by dannyroberts** [comment]
Um, I think since I feel this ballooning a bit, I'm going to go ahead with this one change well tested, and then contemplate the further optimization of lazy fetching as well. (Probably punting the larger refactor unless resourced.)

**Comment by czue** [comment]
that all sounds right. didn't intend to imply we should change that - just wanted to be sure i understood the behavior. thanks for clarifying!

**Comment by benrudolph** [comment]
nice

**Comment by dannyroberts** [comment]
Was a little tricky to trigger the rebuild condition and mock, but finally got it failing on master and passing on this branch!

**Comment by dannyroberts** [comment]
I think good to review and merge once build passes

File: corehq/ex-submodules/casexml/apps/case/tests/test_multimedia.py [hunk 1]
```diff
@@ -230,6 +237,34 @@ def testAttachInUpdate(self, new_attachments=['commcare_logo_file', 'dimagi_logo
             self.assertTrue(attach_name in case.case_attachments)
             self.assertEqual(self._calc_file_hash(attach_name), hashlib.md5(case.get_attachment(attach_name)).hexdigest())
 
+    def testUpdateWithNoNewAttachment(self):
```
Comment (benrudolph) [inline corehq/ex-submodules/casexml/apps/case/tests/test_multimedia.py]: nice, i spent a while trying to reproduce this in tests

**Comment by dannyroberts** [comment]
@czue self-merging for today's deploy

**Comment by czue** [comment]
:+1: thanks!

**Pull request merged**